merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+        for (i, leaf) in self.rightmost_proof.iter_mut().enumerate() {
+            *leaf = H::zero_bytes()[i];
+        }
+    }
+
+    pub fn root(&self) -> [u8; 32] {
+        self.roots[self.current_index as usize]
+    }
+
+    fn inc_current_index(&mut self) {
+        self.current_index = (self.current_index + 1) % MAX_ROOTS as u64;
+    }
+
+    fn update_path_to_leaf(
+        &mut self,
+        leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let mut leaf = *leaf;
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+
+        for (j, sibling) in proof.iter().enumerate() {
+            changelog_path[j] = leaf;
+            leaf = compute_parent_node::<H>(&leaf, sibling, leaf_index, j)?;
+        }
+
+        let changelog = ChangelogEntry::new(leaf, changelog_path, leaf_index);
+        self.inc_current_index();
+        self.changelog[self.current_index as usize] = changelog;
+        self.roots[self.current_index as usize] = leaf;
+
+        Ok(())
+    }
+
+    fn update_inner(
+        &mut self,
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        root_index: usize,
+        proof: &[[u8; 32]; MAX_HEIGHT],
+    ) -> Result<(), HasherError> {
+        let mut updated_leaf = *old_leaf;
+        let mut updated_proof = *proof;
+        for k in root_index..(self.current_index as usize) + 1 {
+            let changelog_entry = self.changelog[k];
+            if leaf_index != changelog_entry.index as usize {
+                // This bit math is used to identify which node in the proof
+                // we need to swap for a corresponding node in a saved change log
+                let padding = 64 - MAX_HEIGHT;
+                let common_path_len = ((leaf_index ^ changelog_entry.index as usize) << padding)
+                    .leading_zeros() as usize;
+                let critbit_index = (MAX_HEIGHT - 1) - common_path_len;
+
+                updated_proof[critbit_index] = changelog_entry.path[critbit_index];
+            } else {
+                updated_leaf = changelog_entry.path[0];
+            }
+        }
+        if is_valid_proof::<H, MAX_HEIGHT>(
+            &self.roots[self.current_index as usize],
+            old_leaf,
+            leaf_index,
+            proof,
+        )? && updated_leaf == *old_leaf
+        {
+            self.update_path_to_leaf(new_leaf, leaf_index, &updated_proof)?
+        }
+
+        Ok(())
+    }
+
+    /// Replaces the `old_leaf` under the `leaf_index` with a `new_leaf`, using
+    /// the given `root` and `proof`.
+    pub fn update(
+        &mut self,
+        root: [u8; 32],
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let proof = full_proof::<H, MAX_HEIGHT>(proof);
+
+        for root_index in 0..(self.current_index as usize) + 1 {
```

Comment:
Why don't we pass in the root_index as a parameter?

Reply:
Also I think there is an edge case here:
``
root_index in 0..(self.current_index as usize) + 1
``
Consider the case that the MAX_ROOT=10, the root index we are looking for is 9, but two Merkle tree updates happend before ours thus current index is 1 -> this loop is not going to find the root because it's only looking in the range 0-2.

Reply:
Ah, damn, I forgot about these points. Will do in next push.